Mathlib/Tactic/Finiteness.lean

```diff
@@ -0,0 +1,36 @@
+/-
+Copyright (c) 2023 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import Mathlib.Tactic.Positivity.Core
+
+/-! # Finiteness tactic
+
+This file implements a basic `finiteness` tactic, designed to solve goals of the form `*** < ∞` and
+(equivalently) `*** ≠ ∞` in the extended nonnegative reals (`ENNReal`, aka `ℝ≥0∞`).
+
+It works recursively according to the syntax of the expression. It is implemented as an `aesop` rule
+set.
+
+[TODO]: improve `finiteness` to also deal with other situations, such as balls in proper spaces with
+a locally finite measure.
+-/
+
+
+open Aesop.BuiltinRules in
+attribute [aesop (rule_sets := [finiteness]) safe -50] assumption intros
```

Comment:
I can understand why you want to avoid `intros`, but why `assumption`?

Reply:
Hmm, not sure. @hrmacbeth, did you have a specific rationale here?

Reply:
I don't remember, but looking now at the `aesop` docs, it seems that the `-50` is a *penalty*, i.e. (since it is negative) actually an incentive to apply these rules.

Reply:
Oh! In that case I need to flip my question - why is `intros` given such a high incentive; it seems like introducing the hypothesis `x = ⊤` would be unhelpful?

Reply:
Very reasonable question :). I see that one of the tests in the PFR version of the tactic addresses this, 
```lean
example {α : Type*} (f : α → ℕ) : ∀ i, (f i : ℝ≥0∞) ≠ ∞ := by finiteness
```
maybe @pitmonticone can report back when he has instituted the test file in this PR.  And also design some other tests which check whether the hypothesis `x = ⊤` is ever being introduced.

Reply:
It doesn't look like that hypothesis is ever being introduced (according to each of my experiments with `set_option trace.aesop true`), although I don't quite understand why - the tactic sets the intro transparency to reducible, and Ne is a reducible def. 
One thing I notice is that
```
example : ∞ ≠ ∞ := by finiteness
```
fails (as it should) but takes quite a while to get there. Is there a way to hard-code in failure (a cut in the search tree, I guess) for something clearly false? @JLimperg 

Here's what the trace looks like:
```
[aesop] ✅️ (G0) [100.0000%] ⋯ ⊢ ⊤ ≠ ⊤ ▶

[aesop] 🆕 ❓️ R0 [100.0000%] safe|tactic|global|Aesop.BuiltinRule.preprocess ▶

[aesop] 🆕 ❓️ G1 [100.0000%] ⋯ ⊢ ⊤ ≠ ⊤ ▶

[aesop] ✅️ (G1) [100.0000%] ⋯ ⊢ ⊤ ≠ ⊤ ▶

[aesop] 🆕 ❓️ R1 [20.0000%] unsafe|apply|global|ne_top_of_lt ▶

[aesop] 🆕 ❓️ G2 [16.0000%] ⋯ ⊢ ⊤ < ?b ▶

[aesop] ✅️ (G2) [16.0000%] ⋯ ⊢ ⊤ < ?b ▶

[aesop] 🆕 ❓️ R2 [18.0000%] safe|apply|global|Ne.lt_top ▶

[aesop] 🆕 ❓️ G3 [18.0000%] ⋯ ⊢ ⊤ ≠ ⊤ ▶

[aesop] ✅️ (G3) [18.0000%] ⋯ ⊢ ⊤ ≠ ⊤ ▶

[aesop] 🆕 ❓️ R3 [3.6000%] unsafe|apply|global|ne_top_of_lt ▶

[aesop] 🆕 ❓️ G4 [2.8800%] ⋯ ⊢ ⊤ < ?h.h.b ▶

[aesop] ✅️ (G4) [2.8800%] ⋯ ⊢ ⊤ < ?h.h.b ▶

[aesop] 🆕 ❓️ R4 [3.2400%] safe|apply|global|Ne.lt_top ▶
```
I also don't know why the success probability goes **up**!

Reply:
> It doesn't look like that hypothesis is ever being introduced (according to each of my experiments with `set_option trace.aesop true`), although I don't quite understand why - the tactic sets the intro transparency to reducible, and Ne is a reducible def.

`Ne` reduces to `Not (. = .)` and `Not` is not reducible, so Aesop's `intros` won't look through it. The `builtin` ruleset contains a rule that explicitly calls `Not.intro`.

Reply:
Ah! That makes perfect sense. I think that clears up all the confusion here, I'll resolve.

Reply:
> Is there a way to hard-code in failure (a cut in the search tree, I guess) for something clearly false? @JLimperg

No, but I think this would be a very nice feature to have! New issue at leanprover-community/aesop#170.

Reply:
> I also don't know why the success probability goes **up**!

This is super weird. I investigated a bit, but haven't found the reason yet.